File: sdk/npm/src/providers/tempmail.ts
import { InternalEmailInfo, Email, Channel } from "../types";
import { normalizeEmail } from "../normalize";
import { fetchWithTimeout } from "../retry";

const CHANNEL: Channel = "tempmail";
const BASE_URL = "https://api.tempmail.ing/api";

const DEFAULT_HEADERS = {
  "User-Agent":
    "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/143.0.0.0 Safari/537.36",
  "Content-Type": "application/json",
  Referer: "https://tempmail.ing/",
  "sec-ch-ua":
    '"Microsoft Edge";v="143", "Chromium";v="143", "Not A(Brand";v="24"',
  "sec-ch-ua-mobile": "?0",
  "sec-ch-ua-platform": '"Windows"',
  DNT: "1",
};

export async function generateEmail(
  duration: number = 30,
): Promise<InternalEmailInfo> {
  const response = await fetchWithTimeout(`${BASE_URL}/generate`, {
    method: "POST",
    headers: DEFAULT_HEADERS,
    body: JSON.stringify({ duration }),
  });

  if (!response.ok) {
    throw new Error(`Failed to generate email: ${response.status}`);
  }

  const data = await response.json();

  if (!data.success) {
    throw new Error("Failed to generate email");
  }

  return {
    channel: CHANNEL,
    email: data.email.address,
    expiresAt: data.email.expiresAt,
    createdAt: data.email.createdAt,
  };
}

export async function getEmails(email: string): Promise<Email[]> {
  const encodedEmail = encodeURIComponent(email);
  const response = await fetchWithTimeout(
    `${BASE_URL}/emails/${encodedEmail}`,
    {
      method: "GET",
      headers: DEFAULT_HEADERS,
    },
  );

  if (!response.ok) {
    throw new Error(`Failed to get emails: ${response.status}`);
  }

  const data = await response.json();

  if (!data.success) {
    throw new Error("Failed to get emails");
  }

  const rawEmails = data.emails || [];
  return rawEmails.map((raw: any) =>
    normalizeEmail(flattenMessage(raw, email), email),
  );
}

/**
 * 将 tempmail.ing 的原始邮件格式扁平化
 *
 * API 返回格式:
 *   from_address: 发件人邮箱
 *   content: HTML 内容（非纯文本）
 *   text: 纯文本（通常为空）
 *   received_at: 接收时间
 *   is_read: 0/1
 */
function flattenMessage(raw: any, recipientEmail: string): any {
  return {
    id: raw.id ?? "",
    from: raw.from_address || raw.from || "",
    to: recipientEmail,
    subject: raw.subject || "",
    text: raw.text || "",
    html: raw.content || raw.html || "",
    date: raw.received_at || raw.date || "",
    isRead: raw.is_read === 1 || raw.is_read === true,
    attachments: raw.attachments || [],
  };
}

File: sdk/go/provider/tempmail.go
package provider

import (
	"bytes"
	"encoding/json"
	"fmt"
	"io"
	"net/url"

	http "github.com/bogdanfinn/fhttp"
)

const tempmailBaseURL = "https://api.tempmail.ing/api"

type TempmailGenerateRequest struct {
	Duration int `json:"duration"`
}

type TempmailGenerateResponse struct {
	Email struct {
		Address         string `json:"address"`
		ExpiresAt       string `json:"expiresAt"`
		CreatedAt       string `json:"createdAt"`
		DurationMinutes int    `json:"durationMinutes"`
	} `json:"email"`
	Success bool `json:"success"`
}

type tempmailEmailsResponse struct {
	Emails  []json.RawMessage `json:"emails"`
	Success bool              `json:"success"`
}

func TempmailGenerate(duration int) (*CreatedMailbox, error) {
	if duration <= 0 {
		duration = 30
	}

	reqBody, _ := json.Marshal(TempmailGenerateRequest{Duration: duration})

	req, err := http.NewRequest("POST", tempmailBaseURL+"/generate", bytes.NewReader(reqBody))
	if err != nil {
		return nil, err
	}

	req.Header.Set("User-Agent", GetCurrentUA())
	req.Header.Set("Content-Type", "application/json")
	req.Header.Set("Referer", "https://tempmail.ing/")
	req.Header.Set("DNT", "1")

	client := HTTPClient()
	resp, err := client.Do(req)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()

	if err := CheckHTTPStatus(resp, "tempmail generate"); err != nil {
		return nil, err
	}

	body, err := io.ReadAll(resp.Body)
	if err != nil {
		return nil, err
	}

	var result TempmailGenerateResponse
	if err := json.Unmarshal(body, &result); err != nil {
		return nil, err
	}

	if !result.Success {
		return nil, fmt.Errorf("failed to generate email")
	}

	info := &CreatedMailbox{Channel: "tempmail", Email: result.Email.Address, Token: ""}
	info.ExpiresAt = result.Email.ExpiresAt
	info.CreatedAt = result.Email.CreatedAt
	return info, nil
}

func TempmailGetEmails(email string) ([]NormEmail, error) {
	encodedEmail := url.QueryEscape(email)

	req, err := http.NewRequest("GET", tempmailBaseURL+"/emails/"+encodedEmail, nil)
	if err != nil {
		return nil, err
	}

	req.Header.Set("User-Agent", GetCurrentUA())
	req.Header.Set("Referer", "https://tempmail.ing/")
	req.Header.Set("DNT", "1")

	client := HTTPClient()
	resp, err := client.Do(req)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()

	if err := CheckHTTPStatus(resp, "tempmail get emails"); err != nil {
		return nil, err
	}

	body, err := io.ReadAll(resp.Body)
	if err != nil {
		return nil, err
	}

	var result tempmailEmailsResponse
	if err := json.Unmarshal(body, &result); err != nil {
		return nil, err
	}

	if !result.Success {
		return nil, fmt.Errorf("failed to get emails")
	}

	/*
	 * tempmail.ing API 返回格式特殊：
	 *   from_address → from
	 *   content → html（是 HTML 内容）
	 *   text → text（通常为空）
	 *   received_at → date
	 *   is_read → 0/1
	 * 需要先扁平化再 normalize
	 */
	emails := make([]NormEmail, 0, len(result.Emails))
	for _, raw := range result.Emails {
		var m map[string]interface{}
		if err := json.Unmarshal(raw, &m); err != nil {
			continue
		}
		flat := tempmailFlattenMessage(m, email)
		emails = append(emails, NormalizeMap(flat, email))
	}
	return emails, nil
}

/*
 * tempmailFlattenMessage 将 tempmail.ing 的原始格式扁平化
 * content 映射到 html，from_address 映射到 from
 */
func tempmailFlattenMessage(raw map[string]interface{}, recipientEmail string) map[string]interface{} {
	return map[string]interface{}{
		"id":      raw["id"],
		"from":    raw["from_address"],
		"to":      recipientEmail,
		"subject": raw["subject"],
		"text":    raw["text"],
		"html":    raw["content"],
		"date":    raw["received_at"],
		"is_read": raw["is_read"],
	}
}
